In [2]:
import pandas as pd
import numpy as np
df = pd.read_csv('int20h-ds-test-dataset.csv')

In [3]:
df

userid user_state  \
0      c95c777785faec8dd910d019d7278ebe         CA   
1      c95c777785faec8dd910d019d7278ebe         CA   
2      f344be2d9a042b7444f3cc5279e38ef1         FL   
3      c95c777785faec8dd910d019d7278ebe         CA   
4      e331ed81422d8fba55520a43a872e701         IL   
...                                 ...        ...   
23352  679eba26c4e75e0afb178360becfa21b         CA   
23353  679eba26c4e75e0afb178360becfa21b         CA   
23354  679eba26c4e75e0afb178360becfa21b         CA   
23355  679eba26c4e75e0afb178360becfa21b         CA   
23356  679eba26c4e75e0afb178360becfa21b         CA   

                       event_name  \
0             Add Vehicle Success   
1               Add Vehicle Break   
2                 Calculator View   
3      Add Payment Method Success   
4                 Sign Up Success   
...                           ...   
23352  Add Payment Method Success   
23353  Account Setup Profile Skip   
23354  Account Setup Profile Skip   
23355    Chat Conversation Opened   
23356        Subscription Premium   

                                        event_attributes   event_created_date  \
0      {"Make":"Dodge","Model":"Caravan","Color":"Whi...  2022-01-16 17:03:04   
1                                                     {}  2022-01-16 17:07:47   
2                                                     {}  2022-01-16 17:16:25   
3         {"Payment Method":"Credit","Tokenized Pay":""}  2022-01-16 17:24:22   
4                                     {"Method":"Apple"}  2022-01-16 17:34:51   
...                                                  ...                  ...   
23352  {"Payment Method":"Credit","Tokenized Pay":"",...  2022-04-16 20:49:24   
23353                               {"Screen":"Address"}  2022-04-16 20:50:05   
23354                          {"Screen":"Phone Number"}  2022-04-16 20:50:10   
23355         {"From":"Dashboard","Transaction type":""}  2022-04-16 20:50:31   
23356  {"Start Date":"2022-04-16","Renewal Date":"202...  2022-04-16 20:50:39   

      event_platform device_manufacture device_model  
0            android            samsung     SM-N975U  
1            android            samsung     SM-N975U  
2            android            samsung    SM-G973U1  
3            android            samsung     SM-N975U  
4                ios              Apple   iPhone12,1  
...              ...                ...          ...  
23352        android             Google     Pixel 3a  
23353        android             Google     Pixel 3a  
23354        android             Google     Pixel 3a  
23355        android             Google     Pixel 3a  
23356      outofband                NaN          NaN  

[23357 rows x 8 columns]

In [4]:
gone_users = df[df['event_name'] == 'Subscription Premium Cancel'].userid.unique().tolist()

In [5]:
df['gone_user'] = df['userid'].isin(gone_users)

In [6]:
count_events = df[['userid','user_state', 'event_name', 'gone_user']]

In [7]:
count_events2 = df[['userid', 'event_name', 'gone_user']]

In [8]:
df3 = count_events2.groupby(['userid', 'event_name']).count().unstack(level=1,fill_value=0)['gone_user']

In [9]:
df3['gone_user'] = df3.index.isin(gone_users)

In [10]:
df4 = df3.groupby(['gone_user']).mean()

In [44]:
df5 = df4.T

In [45]:
df5

gone_user                               False     True 
event_name                                             
Account History Transaction Details  0.323199  2.175299
Account Setup Profile Skip           0.134387  0.223108
Account Setup Skip                   0.064457  0.039841
Add Payment Method Failed            0.093341  0.107570
Add Payment Method Success           0.238978  1.003984
Add Vehicle Break                    0.131043  0.219124
Add Vehicle Failed                   0.006081  0.003984
Add Vehicle Success                  0.514442  0.920319
Calculator Used                      0.030404  0.079681
Calculator View                      0.150198  0.541833
Chat Conversation Opened             0.226817  2.944223
Chat Conversation Started            0.153846  2.772908
Email Confirmation Success           0.238370  0.274900
Order                                1.647613  2.776892
Reset Password Set                   0.000304  0.000000
Sign Out                             0.103071  1.019920
Sign Up Error                        0.007601  0.003984
Sign Up Success                      0.944968  0.880478
Subscription Premium                 0.148373  0.888446
Subscription Premium Cancel          0.000000  1.079681
Subscription Premium Renew           0.083308  0.143426
Transaction Refund                   0.001520  0.386454
Wallet Opened                        0.292186  2.047809

In [107]:
df[df['event_name'] == 'Chat Conversation Started'].shape

(1202, 9)

In [59]:
df.head(2)

userid user_state           event_name  \
0  c95c777785faec8dd910d019d7278ebe         CA  Add Vehicle Success   
1  c95c777785faec8dd910d019d7278ebe         CA    Add Vehicle Break   

                                    event_attributes   event_created_date  \
0  {"Make":"Dodge","Model":"Caravan","Color":"Whi...  2022-01-16 17:03:04   
1                                                 {}  2022-01-16 17:07:47   

  event_platform device_manufacture device_model  gone_user  
0        android            samsung     SM-N975U      False  
1        android            samsung     SM-N975U      False

In [65]:
df.head(2)

userid user_state           event_name  \
0  c95c777785faec8dd910d019d7278ebe         CA  Add Vehicle Success   
1  c95c777785faec8dd910d019d7278ebe         CA    Add Vehicle Break   

                                    event_attributes   event_created_date  \
0  {"Make":"Dodge","Model":"Caravan","Color":"Whi...  2022-01-16 17:03:04   
1                                                 {}  2022-01-16 17:07:47   

  event_platform device_manufacture device_model  gone_user  
0        android            samsung     SM-N975U      False  
1        android            samsung     SM-N975U      False

In [111]:
df_chat = df[df['event_name'] == 'Chat Conversation Opened']
df_chat.shape

(1485, 9)

### 1485 открыли чат тех. поддержки, но лишь 1202 написали тех. поддержке


In [106]:
df.groupby(['device_manufacture']).count()

userid  user_state  event_name  event_attributes  \
device_manufacture                                                     
Alcatel                  1           1           1                 1   
Apple                 9586        9529        9586              9586   
BLU                     20          20          20                20   
FIH                     17          17          17                17   
Google                 361         360         361               361   
HMD Global             115         115         115               115   
HTC                      1           1           1                 1   
HUAWEI                   8           8           8                 8   
KYOCERA                  1           1           1                 1   
LGE                    333         333         333               333   
OPPO                     1           1           1                 1   
OnePlus                171         171         171               171   
TCL                     56          55          56                56   
Umx                      7           7           7                 7   
Wingtech                99          99          99                99   
Xiaomi                   8           8           8                 8   
Yulong                   6           6           6                 6   
ZTE                      2           2           2                 2   
motorola               509         505         509               509   
realme                   6           6           6                 6   
samsung               4527        4519        4527              4527   
vsmart                  12          12          12                12   

                    event_created_date  event_platform  device_model  \
device_manufacture                                                     
Alcatel                              1               1             1   
Apple                             9586            9586          9586   
BLU                                 20              20            20   
FIH                                 17              17            17   
Google                             361             361           361   
HMD Global                         115             115           115   
HTC                                  1               1             1   
HUAWEI                               8               8             8   
KYOCERA                              1               1             1   
LGE                                333             333           333   
OPPO                                 1               1             1   
OnePlus                            171             171           171   
TCL                                 56              56            56   
Umx                                  7               7             7   
Wingtech                            99              99            99   
Xiaomi                               8               8             8   
Yulong                               6               6             6   
ZTE                                  2               2             2   
motorola                           509             509           509   
realme                               6               6             6   
samsung                           4527            4527          4527   
vsmart                              12              12            12   

                    gone_user  
device_manufacture             
Alcatel                     1  
Apple                    9586  
BLU                        20  
FIH                        17  
Google                    361  
HMD Global                115  
HTC                         1  
HUAWEI                      8  
KYOCERA                     1  
LGE                       333  
OPPO                        1  
OnePlus                   171  
TCL                        56  
Umx                         7  
Wingtech                   99  
Xiaomi            

In [104]:
df_chat.drop_duplicates(subset = 'userid').groupby('event_platform').count()

userid  user_state  event_name  event_attributes  \
event_platform                                                     
android            148         148         148               148   
ios                323         323         323               323   

                event_created_date  device_manufacture  device_model  \
event_platform                                                         
android                        148                 148           148   
ios                            323                 323           323   

                gone_user  
event_platform             
android               148  
ios                   323

In [105]:
df_chat.groupby('event_platform').count()

userid  user_state  event_name  event_attributes  \
event_platform                                                     
android            403         403         403               403   
ios               1082        1082        1082              1082   

                event_created_date  device_manufacture  device_model  \
event_platform                                                         
android                        403                 403           403   
ios                           1082                1082          1082   

                gone_user  
event_platform             
android               403  
ios                  1082

In [48]:
df5.insert(2,'mod',df5[True]-df5[False])

In [57]:
df5 = df5[df5['mod']>=0]
df5

gone_user                               False      True       mod
event_name                                                       
Account History Transaction Details  0.323199  2.175299  1.852100
Account Setup Profile Skip           0.134387  0.223108  0.088720
Add Payment Method Failed            0.093341  0.107570  0.014228
Add Payment Method Success           0.238978  1.003984  0.765006
Add Vehicle Break                    0.131043  0.219124  0.088081
Add Vehicle Success                  0.514442  0.920319  0.405877
Calculator Used                      0.030404  0.079681  0.049277
Calculator View                      0.150198  0.541833  0.391635
Chat Conversation Opened             0.226817  2.944223  2.717406
Chat Conversation Started            0.153846  2.772908  2.619062
Email Confirmation Success           0.238370  0.274900  0.036530
Order                                1.647613  2.776892  1.129279
Sign Out                             0.103071  1.019920  0.916849
Subscription Premium                 0.148373  0.888446  0.740073
Subscription Premium Cancel          0.000000  1.079681  1.079681
Subscription Premium Renew           0.083308  0.143426  0.060118
Transaction Refund                   0.001520  0.386454  0.384934
Wallet Opened                        0.292186  2.047809  1.755623

In [ ]:
df1 = count_events.groupby(['user_state','userid', 'event_name']).count().unstack(level=2,fill_value=0)['gone_user']

In [ ]:
df1.reset_index(level = 'user_state', inplace= True)

In [ ]:
df1['gone_user'] = df1.index.isin(gone_users)

In [ ]:
df2 = df1.groupby(['user_state','gone_user']).mean()

In [ ]:
df2

event_name            Account History Transaction Details  \
user_state gone_user                                        
AK         False                                 0.000000   
AL         False                                 0.200000   
           True                                  1.000000   
AR         False                                 0.000000   
           True                                  1.500000   
...                                                   ...   
WA         False                                 0.000000   
WI         False                                 0.117647   
           True                                  0.000000   
WV         False                                 0.000000   
           True                                  3.000000   

event_name            Account Setup Profile Skip  Account Setup Skip  \
user_state gone_user                                                   
AK         False                             0.0            0.000000   
AL         False                             0.0            0.000000   
           True                              0.0            0.000000   
AR         False                             0.0            0.100000   
           True                              0.0            0.000000   
...                                          ...                 ...   
WA         False                             0.0            0.000000   
WI         False                             0.0            0.058824   
           True                              0.0            0.000000   
WV         False                             0.0            0.000000   
           True                              0.0            0.000000   

event_name            Add Payment Method Failed  Add Payment Method Success  \
user_state gone_user                                                          
AK         False                       0.000000                    0.000000   
AL         False                       0.200000                    0.200000   
           True                        0.000000                    1.000000   
AR         False                       0.000000                    0.000000   
           True                        0.000000                    1.000000   
...                                         ...                         ...   
WA         False                       0.000000                    0.000000   
WI         False                       0.176471                    0.294118   
           True                        0.000000                    0.000000   
WV         False                       0.166667                    0.000000   
           True                        0.000000                    1.000000   

event_name            Add Vehicle Break  Add Vehicle Failed  \
user_state gone_user                                          
AK         False               0.000000                 0.0   
AL         False               0.000000                 0.0   
           True                0.000000                 0.0   
AR         False               0.100000                 0.0   
           True                0.000000                 0.0   
...                                 ...                 ...   
WA         False               0.000000                 0.0   
WI         False               0.117647                 0.0   
           True                0.000000                 0.0   
WV         False               0.166667                 0.0   
           True                0.000000                 0.0   

event_name            Add Vehicle Success  Calculator Used  Calculator View  \
user_state gone_user                                                          
AK         False                 0.000000         0.000000         0.000000   
AL         False                 0.400000         0.000000         0.000000   
           True                  1.000000         0.000000         0.000000   
AR         False          